File: python/tests/test_gmod.py
import unittest
from dataclasses import dataclass
from typing import List

from vista_sdk.Client import Client
from vista_sdk.Gmod import Gmod, TraversalHandlerResult, TraversalOptions
from vista_sdk.GmodDto import GmodDto
from vista_sdk.GmodNode import GmodNode
from vista_sdk.GmodPath import GmodPath
from vista_sdk.VIS import VIS
from vista_sdk.VisVersions import VisVersion, VisVersionExtension

from .test_vis import TestVis


class TestGmod(unittest.TestCase):
    def setUp(self):
        self.vis = TestVis.get_vis()
        Client.get_gmod_test(VisVersionExtension.to_version_string(VisVersion.v3_4a))

    def test_gmod_loads(self):
        for version in VisVersion:
            with self.subTest(version=version):
                gmod = self.vis.get_gmod(version)
                self.assertIsNotNone(gmod)
                self.assertTrue(gmod.try_get_node("400a"))

    def test_gmod_properties(self):
        for version in VisVersion:
            with self.subTest(version=version):
                gmod = self.vis.get_gmod(version)
                self.assertIsNotNone(gmod)

                nodes = list(gmod._node_map)
                min_length = min(gmod, key=lambda x: len(x[1].code))
                max_length = max(gmod, key=lambda x: len(x[1].code))

                self.assertEqual(2, len(min_length[1].code))
                self.assertEqual("VE", min_length[1].code)
                self.assertEqual(10, len(max_length[1].code))
                possible_max = ["C1053.3111", "H346.11113"]
                self.assertIn(max_length[1].code, possible_max)

                expected_counts = [6420, 6557, 6672]
                self.assertIn(len(nodes), expected_counts)

    def test_gmod_lookup(self):
        for version in VisVersion:
            with self.subTest(version=version):
                gmod = self.vis.get_gmod(version)
                self.assertIsNotNone(gmod)

                gmodDto: GmodDto = self.vis.get_gmod_dto(version)
                self.assertIsNotNone(gmodDto)

                seen = set()
                for node in gmodDto.items:
                    self.assertIsNotNone(node.code)
                    self.assertTrue(node.code not in seen, f"code: {node.code}")
                    seen.add(node.code)

                    success, foundNode = gmod.try_get_node(node.code)
                    self.assertTrue(success)
                    self.assertIsNotNone(foundNode)
                    if foundNode is not None:
                        self.assertEqual(node.code, foundNode.code)

                seen.clear()
                counter = 0
                for node in gmod:
                    self.assertIsNotNone(node[1].code)
                    self.assertTrue(node[1].code not in seen, f"code: {node[1].code}")
                    seen.add(node[1].code)

                    success, foundNode = gmod.try_get_node(node[1].code)
                    self.assertTrue(success)
                    self.assertIsNotNone(foundNode)
                    if foundNode is not None:
                        self.assertEqual(node[1].code, foundNode.code)
                    counter += 1

                self.assertEqual(len(gmodDto.items), counter)

                test_codes = [
                    "ABC",
                    None,
                    "",
                    "SDFASDFSDAFb",
                    "✅",
                    "a✅b",
                    "ac✅bc",
                    "✅bc",
                    "a✅",
                    "ag✅",
                ]
                for code in test_codes:
                    success, _ = gmod.try_get_node(code)
                    self.assertFalse(success)

    def test_gmod_node_equality(self):
        gmod = self.vis.get_gmod(VisVersion.v3_4a)

        node1 = gmod["400a"]
        node2 = gmod["400a"]

        self.assertEqual(node1, node2, "Nodes should be equal")
        self.assertIs(node1, node2, "Nodes should be the exact same object (identity)")

        if node2 is not None:
            node3 = node2.with_location("1")
        self.assertNotEqual(
            node1, node3, "Nodes should not be equal after modification"
        )
        self.assertIsNot(
            node1, node3, "Nodes should not be the same object after modification"
        )

        if node2 is not None:
            node4 = node2.clone()
        self.assertEqual(node1, node4, "Cloned nodes should be equal")
        self.assertIsNot(node1, node4, "Cloned nodes should not be the same object")

    def test_gmod_node_types(self):
        gmod = self.vis.get_gmod(VisVersion.v3_4a)

        unique_types = set()
        for node in gmod:
            category_type = f"{node[1].metadata.category} | {node[1].metadata.type}"
            unique_types.add(category_type)

        self.assertTrue(unique_types, "The set of node types should not be empty")

    def test_gmod_root_node_children(self):
        for version in VisVersion:
            with self.subTest(version=version):
                gmod = self.vis.get_gmod(version)

                node = gmod.root_node
                if node is None:
                    self.fail("Root node should not be None")
                else:
                    self.assertIsNotNone(node, "Root node should not be None")
                    self.assertGreater(
                        len(node.children), 0, "Root node should have children"
                    )

    def test_normal_assignments(self):
        gmod = self.vis.get_gmod(VisVersion.v3_4a)

        node = gmod["411.3"]
        self.assertIsNotNone(
            node.product_type, "ProductType should not be None for node '411.3'"
        )
        self.assertIsNone(
            node.product_selection, "ProductSelection should be None for node '411.3'"
        )

        node = gmod["H601"]
        self.assertIsNone(
            node.product_type, "ProductType should be None for node 'H601'"
        )

    def test_node_with_product_selection(self):
        gmod = self.vis.get_gmod(VisVersion.v3_4a)

        node = gmod["411.2"]
        self.assertIsNotNone(
            node.product_selection,
            "ProductSelection should not be None for node '411.2'",
        )
        self.assertIsNone(
            node.product_type, "ProductType should be None for node '411.2'"
        )

        node = gmod["H601"]
        self.assertIsNone(
            node.product_selection, "ProductSelection should be None for node 'H601'"
        )

    def test_product_selection(self):
        gmod = self.vis.get_gmod(VisVersion.v3_4a)

        node = gmod["CS1"]
        self.assertTrue(
            node.is_product_selection,
            "Node 'CS1' should be identified as having a product selection",
        )

    def test_mappability(self):
        gmod = self.vis.get_gmod(VisVersion.v3_4a)

        test_cases = [
            ("VE", False),
            ("300a", False),
            ("300", True),
            ("411", True),
            ("410", True),
            ("651.21s", False),
            ("924.2", True),
            ("411.1", False),
            ("C101", True),
            ("CS1", False),
            ("C101.663", True),
            ("C101.4", True),
            ("C101.21s", False),
            ("F201.11", True),
            ("C101.211", False),
        ]

        for code, expected_mappable in test_cases:
            with self.subTest(code=code, mappable=expected_mappable):
                node = gmod[code]
                self.assertEqual(
                    expected_mappable,
                    node.is_mappable,
                    f"Node {code} mappability should be {expected_mappable}",
                )

    def occurrences(self, parents: List[GmodNode], node: GmodNode) -> int:
        count = 0
        for parent in parents:
            if parent.code == node.code:
                count += 1
        return count

    def test_full_traversal(self):
        gmod = VIS().get_gmod(VisVersion.v3_4a)

        max_expected = TraversalOptions.DEFAULT_MAX_TRAVERSAL_OCCURRENCE

        class State:
            count: int
            max: int
            paths: List[GmodPath]

            def __init__(self):
                self.count = 0
                self.max = 0
                self.paths = []

            def increment(self):
                self.count += 1

        def traversal_handler(state: State, parents: List[GmodNode], node: GmodNode):
            self.assertTrue(
                len(parents) == 0 or parents[0].is_root(),
                "First parent should be root or no parents",
            )

            state.increment()

            def sample_test(parents: List[GmodNode], node: GmodNode) -> bool:
                if node.code == "HG3":
                    return True
                return any(parent.code == "HG3" for parent in parents)

            if sample_test(parents, node):
                state.paths.append(GmodPath(parents[:], node))

            skip_occurrence_check = Gmod.is_product_selection_assignment(
                parents[-1] if len(parents) > 0 else None, node
            )
            if skip_occurrence_check:
                return TraversalHandlerResult.CONTINUE

            occ = self.occurrences(parents, node)
            if occ > state.max:
                state.max = occ

            return TraversalHandlerResult.CONTINUE

        state: State = State()

        completed = gmod.traverse(args1=state, args2=traversal_handler)
        self.assertEqual(
            max_expected, state.max, "Maximum occurrence should match expected"
        )
        self.assertTrue(completed, "Traversal should complete successfully")

    @unittest.skip("This test is too slow to run in CI")
    def test_full_traversal_with_options(self):
        gmod = self.vis.get_gmod(VisVersion.v3_4a)

        max_expected = 2
        max_occurrence = 0

        def traversal_handler(parents, node):
            skip_occurrence_check = Gmod.is_product_selection_assignment(
                parents[-1] if parents else None, node
            )
            if skip_occurrence_check:
                return TraversalHandlerResult.CONTINUE

            occ = self.occurrences(parents, node)
            nonlocal max_occurrence
            if occ > max_occurrence:
                max_occurrence = occ

            return TraversalHandlerResult.CONTINUE

        options = TraversalOptions(max_traversal_occurrence=max_expected)
        completed = gmod.traverse(args1=traversal_handler, args2=options)

        self.assertEqual(
            max_expected,
            max_occurrence,
            "Maximum occurrence should match the expected limit",
        )
        self.assertTrue(completed, "Traversal should complete successfully")

    def test_partial_traversal(self):
        gmod = self.vis.get_gmod(VisVersion.v3_4a)

        state = self.TraversalState(stop_after=5)

        def traversal_handler(state, parents, node):
            self.assertTrue(
                len(parents) == 0 or parents[0].is_root,
                "First parent should be root or no parents",
            )
            state.node_count += 1
            if state.node_count == state.stop_after:
                return TraversalHandlerResult.STOP
            return TraversalHandlerResult.CONTINUE

        completed = gmod.traverse(args1=state, args2=traversal_handler)

        self.assertEqual(
            state.stop_after,
            state.node_count,
            "Traversal should stop after the specified number of nodes",
        )
        self.assertFalse(
            completed, "Traversal should not complete fully (should stop early)"
        )

    def test_full_traversal_from(self):
        gmod = self.vis.get_gmod(VisVersion.v3_4a)

        state = self.TraversalState(stop_after=0)

        def traversal_handler(state, parents, node):
            self.assertTrue(
                len(parents) == 0 or parents[0].code == "400a",
                "First parent should be root or no parents",
            )
            state.node_count += 1
            return TraversalHandlerResult.CONTINUE

        completed = gmod.traverse(
            args1=state, args2=gmod["400a"], args3=traversal_handler
        )

        self.assertTrue(completed, "Traversal should complete full")

    @dataclass
    class TraversalState:
        stop_after: int
        node_count: int = 0


if __name__ == "__main__":
    unittest.main()

File: cpp/test/TEST_Gmod.cpp
/**
 * @file GmodTests.cpp
 * @brief Unit tests for the Generic Product Model (GMOD) functionality
 */

#include "pch.h"

#include "dnv/vista/sdk/GmodTraversal.h"
#include "dnv/vista/sdk/GmodPath.h"
#include "dnv/vista/sdk/VIS.h"

namespace dnv::vista::sdk::tests
{
	namespace GmodTestsFixture
	{
		class GmodTests : public ::testing::Test
		{
		public:
			static std::pair<VIS&, const Gmod&> visAndGmod( VisVersion visVersion )
			{
				VIS& vis = VIS::instance();
				const auto& gmod = vis.gmod( visVersion );

				return { vis, gmod };
			}
		};

		//=====================================================================
		// Helper functions
		//=====================================================================

		static std::size_t occurrences( const std::vector<const GmodNode*>& parents, const GmodNode& node )
		{
			return static_cast<size_t>(
				std::count_if(
					parents.begin(), parents.end(),
					[&node]( const GmodNode* p ) {
						return p != nullptr && p->code() == node.code();
					} ) );
		}

		//=====================================================================
		// Traversal Helper Struct
		//=====================================================================

		struct TraversalState
		{
			TraversalState( int stopAfter )
				: stopAfter{ stopAfter },
				  nodeCount{ 0 }
			{
			}

			int stopAfter;
			int nodeCount;
		};

		//=====================================================================
		// Tests
		//=====================================================================

		//----------------------------------------------
		// Test_Gmod_Node_Equality
		//----------------------------------------------

		TEST_F( GmodTests, Test_Gmod_Node_Equality )
		{
			auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

			const GmodNode& node1 = gmod["400a"];
			const GmodNode& node2 = gmod["400a"];

			EXPECT_EQ( node1, node2 );
			EXPECT_EQ( &node1, &node2 );

			GmodNode node3 = node2.withLocation( "1" );
			EXPECT_NE( node1, node3 );
			EXPECT_NE( &node1, &node3 );

			GmodNode node4 = node2;
			EXPECT_EQ( node1, node4 );
			EXPECT_NE( &node1, &node4 );
		}

		//----------------------------------------------
		// Test_Gmod_Node_Types
		//----------------------------------------------

		TEST_F( GmodTests, Test_Gmod_Node_Types )
		{
			auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

			std::unordered_set<std::string> set;
			Gmod::Enumerator enumerator = gmod.enumerator();

			while ( enumerator.next() )
			{
				const GmodNode& node = enumerator.current();
				const GmodNodeMetadata& metadata = node.metadata();
				std::string category_type_string = metadata.category();
				category_type_string += " | ";
				category_type_string += metadata.type();
				set.insert( category_type_string );
			}

			EXPECT_FALSE( set.empty() );
		}

		//----------------------------------------------
		// Test_Normal_Assignments
		//----------------------------------------------

		TEST_F( GmodTests, Test_Normal_Assignments )
		{
			auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

			const GmodNode& node1 = gmod["411.3"];
			EXPECT_NE( node1.productType(), nullptr );
			EXPECT_EQ( node1.productSelection(), nullptr );

			const GmodNode& node2 = gmod["H601"];
			EXPECT_EQ( node2.productType(), nullptr );
		}

		//----------------------------------------------
		// Test_Node_With_Product_Selection
		//----------------------------------------------

		TEST_F( GmodTests, Test_Node_With_Product_Selection )
		{
			auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

			const GmodNode& node1 = gmod["411.2"];
			EXPECT_NE( node1.productSelection(), nullptr );
			EXPECT_EQ( node1.productType(), nullptr );

			const GmodNode& node2 = gmod["H601"];
			EXPECT_EQ( node2.productSelection(), nullptr );
		}

		//----------------------------------------------
		// Test_Product_Selection
		//----------------------------------------------

		TEST_F( GmodTests, Test_Product_Selection )
		{
			auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

			const GmodNode& node = gmod["CS1"];
			EXPECT_TRUE( node.isProductSelection() );
		}

		//----------------------------------------------
		// Test_Full_Traversal
		//----------------------------------------------

		TEST_F( GmodTests, Test_Full_Traversal )
		{
			auto visVersionForPath = VisVersion::v3_4a;
			auto [vis, gmod] = visAndGmod( visVersionForPath );

			struct FullTraversalState
			{
				std::vector<GmodPath> paths;
				int maxOccurrence = 0;
				const Gmod& gmodRef;

				FullTraversalState( const Gmod& gmod ) : gmodRef( gmod ) {}
			};

			FullTraversalState traversalState( gmod );
			int maxExpected = TraversalOptions::DEFAULT_MAX_TRAVERSAL_OCCURRENCE;

			TraverseHandlerWithState<FullTraversalState> handler =
				[]( FullTraversalState& state, const std::vector<const GmodNode*>& parents, const GmodNode& node ) -> TraversalHandlerResult {
				EXPECT_TRUE( parents.empty() || parents[0]->isRoot() );

				bool isHG3Related = ( node.code() == "HG3" ) ||
									std::any_of( parents.begin(), parents.end(), []( const GmodNode* p ) {
										return p != nullptr && p->code() == "HG3";
									} );

				if ( isHG3Related )
				{
					std::vector<GmodNode*> nonConstParents;
					nonConstParents.reserve( parents.size() );
					for ( const GmodNode* p_const : parents )
					{
						nonConstParents.push_back( const_cast<GmodNode*>( p_const ) );
					}
					GmodNode* nonConstNode = const_cast<GmodNode*>( &node );

					state.paths.emplace_back( state.gmodRef, nonConstNode, std::move( nonConstParents ) );
				}

				const GmodNode* lastParent = parents.empty() ? nullptr : parents.back();

				bool skipOccurenceCheck = Gmod::isProductSelectionAssignment( lastParent, &node );
				if ( skipOccurenceCheck )
				{
					return TraversalHandlerResult::Continue;
				}

				int occ = static_cast<int>( occurrences( parents, node ) );
				if ( occ > state.maxOccurrence )
				{
					state.maxOccurrence = occ;
				}

				return TraversalHandlerResult::Continue;
			};

			bool completed = GmodTraversal::traverse( traversalState, gmod, handler );

			EXPECT_EQ( maxExpected, traversalState.maxOccurrence );
			EXPECT_TRUE( completed );
		}

		//----------------------------------------------
		// Test_Full_Traversal_With_Options
		//----------------------------------------------

		TEST_F( GmodTests, Test_Full_Traversal_With_Options )
		{
			auto visVersionForPath = VisVersion::v3_4a;
			auto [vis, gmod] = visAndGmod( visVersionForPath );

			struct MaxOccurrenceState
			{
				int maxOccurrence = 0;
			};

			MaxOccurrenceState state;
			size_t maxExpected = 2;
			TraversalOptions options;
			options.maxTraversalOccurrence = maxExpected;

			TraverseHandlerWithState<MaxOccurrenceState> handler =
				[]( MaxOccurrenceState& state, const std::vector<const GmodNode*>& parents, const GmodNode& node ) -> TraversalHandlerResult {
				const GmodNode* lastParent = parents.empty() ? nullptr : parents.back();
				bool skipOccurenceCheck = Gmod::isProductSelectionAssignment( lastParent, &node );
				if ( skipOccurenceCheck )
				{
					return TraversalHandlerResult::Continue;
				}

				int occ = static_cast<int>( occurrences( parents, node ) );
				if ( occ > state.maxOccurrence )
				{
					state.maxOccurrence = occ;
				}

				return TraversalHandlerResult::Continue;
			};

			bool completed = GmodTraversal::traverse( state, gmod, handler, options );

			EXPECT_EQ( maxExpected, state.maxOccurrence );
			EXPECT_TRUE( completed );
		}

		//----------------------------------------------
		// Test_Partial_Traversal
		//----------------------------------------------

		TEST_F( GmodTests, Test_Partial_Traversal )
		{
			auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

			TraversalState state( 5 );

			dnv::vista::sdk::TraverseHandlerWithState<TraversalState> handler =
				[]( TraversalState& currentState, const std::vector<const GmodNode*>& parents, [[maybe_unused]] const GmodNode& node ) -> TraversalHandlerResult {
				EXPECT_TRUE( parents.empty() || parents[0]->isRoot() );
				if ( ++currentState.nodeCount == currentState.stopAfter )
				{
					return TraversalHandlerResult::Stop;
				}

				return TraversalHandlerResult::Continue;
			};

			bool completed = GmodTraversal::traverse( state, gmod, handler );

			EXPECT_EQ( state.stopAfter, state.nodeCount );
			EXPECT_FALSE( completed );
		}

		//----------------------------------------------
		// Test_Full_Traversal_From
		//----------------------------------------------

		TEST_F( GmodTests, Test_Full_Traversal_From )
		{
			auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

			TraversalState state( 0 );

			const GmodNode& startNode = gmod["400a"];

			dnv::vista::sdk::TraverseHandlerWithState<TraversalState> handler =
				[]( TraversalState& currentState, const std::vector<const GmodNode*>& parents, [[maybe_unused]] const GmodNode& node ) -> TraversalHandlerResult {
				EXPECT_TRUE( parents.empty() || ( parents[0] != nullptr && parents[0]->code() == "400a" ) );

				++currentState.nodeCount;

				return TraversalHandlerResult::Continue;
			};

			bool completed = GmodTraversal::traverse( state, startNode, handler );

			EXPECT_TRUE( completed );
		}
	}

	namespace GmodTestsParametrized
	{
		class GmodTests : public ::testing::TestWithParam<VisVersion>
		{
		public:
			static std::pair<VIS&, const Gmod&> visAndGmod( VisVersion visVersion )
			{
				VIS& vis = VIS::instance();
				const auto& gmod = vis.gmod( visVersion );

				return { vis, gmod };
			}
		};

		//=====================================================================
		// Tests
		//=====================================================================

		//----------------------------------------------
		// Test_Gmod_Loads
		//----------------------------------------------

		TEST_P( GmodTests, Test_Gmod_Loads )
		{
			auto visVersion = GetParam();
			auto [vis, gmod] = visAndGmod( visVersion );

			const GmodNode* tempNodePtr = nullptr;
			ASSERT_TRUE( gmod.tryGetNode( std::string( "400a" ), tempNodePtr ) ) << "Node '400a' not found in GMOD for version " << VisVersionExtensions::toVersionString( visVersion );
			ASSERT_NE( tempNodePtr, nullptr ) << "tryGetNode succeeded but pointer is null for '400a' in GMOD version " << VisVersionExtensions::toVersionString( visVersion );
		}

		//----------------------------------------------
		// Test_Gmod_Properties
		//----------------------------------------------

		struct ExpectedValues
		{
			std::string maxCode;
			int nodeCount;
		};

		const std::map<VisVersion, ExpectedValues> expectedMaxes = {
			{ VisVersion::v3_4a, { "C1053.3114", 6420 } },
			{ VisVersion::v3_5a, { "C1053.3114", 6557 } },
			{ VisVersion::v3_6a, { "C1053.3114", 6557 } },
			{ VisVersion::v3_7a, { "H346.11113", 6672 } },
			{ VisVersion::v3_8a, { "H346.11113", 6335 } } };

		TEST_P( GmodTests, Test_Gmod_Properties )
		{
			/*
			 * This test ensures certain properties of the Gmod data
			 * that we make some design desicisions based on,
			 * i.e for hashing of the node code
			 */
			auto visVersion = GetParam();
			auto [vis, gmod] = visAndGmod( visVersion );

			const GmodNode* minLengthLexiographicallyOrderedNode = nullptr;
			const GmodNode* maxLengthLexiographicallyOrderedNode = nullptr;
			size_t currentMinLength = std::string::npos;
			size_t currentMaxLength = 0;
			int nodeCount = 0;

			Gmod::Enumerator enumerator = gmod.enumerator();
			while ( enumerator.next() )
			{
				const GmodNode& node = enumerator.current();
				nodeCount++;
				const std::string& code = node.code();
				size_t len = code.length();

				if ( !minLengthLexiographicallyOrderedNode || len < currentMinLength )
				{
					currentMinLength = len;
					minLengthLexiographicallyOrderedNode = &node;
				}
				else if ( len == currentMinLength )
				{
					if ( minLengthLexiographicallyOrderedNode && code < minLengthLexiographicallyOrderedNode->code() )
					{
						minLengthLexiographicallyOrderedNode = &node;
					}
				}

				if ( !maxLengthLexiographicallyOrderedNode || len > currentMaxLength )
				{
					currentMaxLength = len;
					maxLengthLexiographicallyOrderedNode = &node;
				}
				else if ( len == currentMaxLength )
				{
					if ( maxLengthLexiographicallyOrderedNode && code > maxLengthLexiographicallyOrderedNode->code() )
					{
						maxLengthLexiographicallyOrderedNode = &node;
					}
				}
			}

			ASSERT_NE( minLengthLexiographicallyOrderedNode, nullptr ) << "minLengthNode should not be null for " << VisVersionExtensions::toVersionString( visVersion );
			ASSERT_NE( maxLengthLexiographicallyOrderedNode, nullptr ) << "maxLengthNode should not be null for " << VisVersionExtensions::toVersionString( visVersion );

			ASSERT_EQ( minLengthLexiographicallyOrderedNode->code().length(), 2 ) << "Min code length mismatch for " << VisVersionExtensions::toVersionString( visVersion );
			ASSERT_EQ( minLengthLexiographicallyOrderedNode->code(), "VE" ) << "Min code value mismatch for " << VisVersionExtensions::toVersionString( visVersion );

			ASSERT_EQ( maxLengthLexiographicallyOrderedNode->code().length(), 10 ) << "Max code length mismatch for " << VisVersionExtensions::toVersionString( visVersion );

			auto expectedIt = expectedMaxes.find( visVersion );
			ASSERT_NE( expectedIt, expectedMaxes.end() ) << "Expected values not found for GMOD version " << VisVersionExtensions::toVersionString( visVersion );
			const auto& expectedValues = expectedIt->second;

			ASSERT_EQ( maxLengthLexiographicallyOrderedNode->code(), expectedValues.maxCode ) << "Max code value mismatch for " << VisVersionExtensions::toVersionString( visVersion );
			ASSERT_EQ( nodeCount, expectedValues.nodeCount ) << "Node count mismatch for " << VisVersionExtensions::toVersionString( visVersion );
		}

		//----------------------------------------------
		// Test_Gmod_Lookup
		//----------------------------------------------

		TEST_P( GmodTests, Test_Gmod_Lookup )
		{
			auto visVersion = GetParam();
			auto [vis, gmod] = visAndGmod( visVersion );

			const GmodDto& gmodDtoObject = vis.gmodDto( visVersion );

			{
				std::unordered_set<std::string> seen_codes;
				for ( const GmodNodeDto& nodeDto : gmodDtoObject.items() )
				{
					const std::string& dtoCode = nodeDto.code();
					ASSERT_FALSE( dtoCode.empty() ) << "DTO code is empty for version " << VisVersionExtensions::toVersionString( visVersion );

					auto insert_result = seen_codes.insert( dtoCode );
					ASSERT_TRUE( insert_result.second ) << "Duplicate DTO code: " << dtoCode << " for version " << VisVersionExtensions::toVersionString( visVersion );

					const GmodNode* foundNodePtr = nullptr;
					ASSERT_TRUE( gmod.tryGetNode( dtoCode, foundNodePtr ) ) << "Failed to find node from DTO code: " << dtoCode << " for version " << VisVersionExtensions::toVersionString( visVersion );
					ASSERT_NE( foundNodePtr, nullptr ) << "Found node pointer is null for DTO code: " << dtoCode << " for version " << VisVersionExtensions::toVersionString( visVersion );
					ASSERT_EQ( dtoCode, foundNodePtr->code() ) << "Mismatch between DTO code and found node code for: " << dtoCode << " for version " << VisVersionExtensions::toVersionString( visVersion );
				}
			}

			int gmodIteratedCount = 0;
			{
				std::unordered_set<std::string> seen_codes;
				Gmod::Enumerator enumerator = gmod.enumerator();
				while ( enumerator.next() )
				{
					const GmodNode& node = enumerator.current();
					const std::string& gmodNodeCode = node.code();
					ASSERT_FALSE( gmodNodeCode.empty() ) << "Gmod iterated node code is empty for version " << VisVersionExtensions::toVersionString( visVersion );

					auto insert_result = seen_codes.insert( gmodNodeCode );
					ASSERT_TRUE( insert_result.second ) << "Duplicate Gmod iterated code: " << gmodNodeCode << " for version " << VisVersionExtensions::toVersionString( visVersion );

					const GmodNode* foundNodePtr = nullptr;
					ASSERT_TRUE( gmod.tryGetNode( gmodNodeCode, foundNodePtr ) ) << "Failed to find node from Gmod iterated code: " << gmodNodeCode << " for version " << VisVersionExtensions::toVersionString( visVersion );
					ASSERT_NE( foundNodePtr, nullptr ) << "Found node pointer is null for Gmod iterated code: " << gmodNodeCode << " for version " << VisVersionExtensions::toVersionString( visVersion );
					ASSERT_EQ( gmodNodeCode, foundNodePtr->code() ) << "Mismatch between Gmod iterated code and found node code for: " << gmodNodeCode << " for version " << VisVersionExtensions::toVersionString( visVersion );
					gmodIteratedCount++;
				}
			}

			const GmodNode* tempNodePtr = nullptr;
			ASSERT_FALSE( gmod.tryGetNode( std::string( "ABC" ), tempNodePtr ) );
			ASSERT_FALSE( gmod.tryGetNode( std::string_view( "" ), tempNodePtr ) );
			ASSERT_FALSE( gmod.tryGetNode( std::string_view( "SDFASDFSDAFb" ), tempNodePtr ) );
			ASSERT_FALSE( gmod.tryGetNode( std::string_view( "✅" ), tempNodePtr ) );
			ASSERT_FALSE( gmod.tryGetNode( std::string_view( "a✅b" ), tempNodePtr ) );
			ASSERT_FALSE( gmod.tryGetNode( std::string_view( "ac✅bc" ), tempNodePtr ) );
			ASSERT_FALSE( gmod.tryGetNode( std::string_view( "✅bc" ), tempNodePtr ) );
			ASSERT_FALSE( gmod.tryGetNode( std::string_view( "a✅" ), tempNodePtr ) );
			ASSERT_FALSE( gmod.tryGetNode( std::string_view( "ag✅" ), tempNodePtr ) );
		}

		//----------------------------------------------
		// Test_Gmod_RootNode_Children
		//----------------------------------------------

		TEST_P( GmodTests, Test_Gmod_RootNode_Children )
		{
			auto visVersion = GetParam();
			auto [vis, gmod] = visAndGmod( visVersion );

			const GmodNode& node = gmod.rootNode();

			EXPECT_FALSE( node.children().empty() );
		}

		//----------------------------------------------
		// Test_Mappability
		//----------------------------------------------

		struct MappabilityTestData
		{
			std::string code;
			bool expectedMappable;
		};

		class GmodMappabilityTests : public ::testing::TestWithParam<MappabilityTestData>
		{
		};

		TEST_P( GmodMappabilityTests, Test_Mappability )
		{
			const auto& testData = GetParam();
			auto [vis, gmod] = GmodTests::visAndGmod( VisVersion::v3_4a );

			const GmodNode& node = gmod[testData.code];
			EXPECT_EQ( node.isMappable(), testData.expectedMappable );
		}

		//=====================================================================
		// Instantiate
		//=====================================================================

		INSTANTIATE_TEST_SUITE_P(
			GmodTestSuite,
			GmodTests,
			::testing::Values(
				VisVersion::v3_4a,
				VisVersion::v3_5a,
				VisVersion::v3_6a,
				VisVersion::v3_7a,
				VisVersion::v3_8a ) );

		INSTANTIATE_TEST_SUITE_P(
			GmodMappabilityTestSuite,
			GmodMappabilityTests,
			::testing::Values(
				MappabilityTestData{ "VE", false },
				MappabilityTestData{ "300a", false },
				MappabilityTestData{ "300", true },
				MappabilityTestData{ "411", true },
				MappabilityTestData{ "410", true },
				MappabilityTestData{ "651.21s", false },
				MappabilityTestData{ "924.2", true },
				MappabilityTestData{ "411.1", false },
				MappabilityTestData{ "C101", true },
				MappabilityTestData{ "CS1", false },
				MappabilityTestData{ "C101.663", true },
				MappabilityTestData{ "C101.4", true },
				MappabilityTestData{ "C101.21s", false },
				MappabilityTestData{ "F201.11", true },
				MappabilityTestData{ "C101.211", false } ) );
	}
}
